### PR TITLE
feat(dataflow): Log if DataFlow declares a block that FxA does not have.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
issue #399

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
To manually test, a GCP account is needed with a Service Account and a PubSub project and topic.

@ameihm0912 [provided this list of DataFlow events](https://gist.github.com/ameihm0912/dc421ba3850e0c1cabed47e29384e654) so that I could send those events through a PubSub subscription and ensure the appropriate handler was used.

The customs server provides a way to [block IP addresses](https://github.com/mozilla/fxa/blob/1cef84f26e3452365cdfda5630cb09396feab15d/packages/fxa-customs-server/lib/server.js#L510) and [block email addresses](https://github.com/mozilla/fxa/blob/1cef84f26e3452365cdfda5630cb09396feab15d/packages/fxa-customs-server/lib/server.js#L481) using HTTP requests. There is not yet a way to use HTTP requests to block a uid or email/IP address combination, and I didn't much feel like adding those methods because it would make the PR blow up even more.

I was able to use this flow to test IP-only blocks (rl_statuscheck, rl_sms_sourceaddress):

1. Ensure a DataFlow event that deals only with IP blocks logs a "dataflow.unexpected_block" message.
2. Call `blockIp` using wget to add a block for the IP address
```bash
wget --post-data='{"ip":"127.0.0.1"}' --header='Content-Type: application/json' http://127.0.0.1:7000/blockIp
```
3. Re-send the event from 1 to ensure the message is received, but no "dataflow.unexpected_block" is logged.

The same process is used to test email-only blocks (rl_sms_accountid, rl_email_recipient) by calling `blockEmail`.

An example of what this flow looks like in the customs-server console for an email block:

### Send DataFlow event over PubSub to ensure message is received.
```json
{"name":"customs-server","hostname":"stomlinson-30887","pid":60809,"level":30,"op":"dataflow.message","count":0,"id":"512790085216109","data":"{\"severity\":\"info\",\"id\":\"250f4821-6920-4832-bfe3-0a7a1bf1c1a6\",\"summary\":\"test sms send account id rate violation, scotty@mozilla.com\",\"category\":\"customs\",\"timestamp\":\"1970-01-01T00:00:00.000Z\",\"metadata\":[{\"key\":\"customs_category\",\"value\":\"rl_sms_accountid\"},{\"key\":\"threshold\",\"value\":\"5\"},{\"key\":\"count\",\"value\":\"5\"},{\"key\":\"notify_merge\",\"value\":\"rl_sms_accountid\"},{\"key\":\"accountid\",\"value\":\"scotty@mozilla.com\"},{\"key\":\"masked_summary\",\"value\":\"test sms send account id rate violation, <<masked>>\"}]}","attributes":{},"msg":"","time":"2019-04-10T11:12:37.680Z","v":0}
```
### email address is not yet blocked, so `dataflow.block.unexpected` is logged
```json
{"name":"customs-server","hostname":"stomlinson-30887","pid":60809,"level":30,"customs_category":"rl_sms_accountid","threshold":"5","count":"5","notify_merge":"rl_sms_accountid","accountid":"scotty@mozilla.com","masked_summary":"test sms send account id rate violation, <<masked>>","op":"dataflow.block.unexpected","msg":"","time":"2019-04-10T11:12:37.689Z","v":0}
```
### Call `/blockEmail` with email address
```json
{"name":"customs-server","hostname":"stomlinson-30887","pid":60809,"level":30,"op":"handleBan.blockEmail","email":"scotty@mozilla.com","msg":"","time":"2019-04-10T11:12:43.378Z","v":0}
```
```json
{"name":"customs-server","hostname":"stomlinson-30887","pid":60809,"level":30,"op":"request.blockEmail","email":"scotty@mozilla.com","msg":"","time":"2019-04-10T11:12:43.381Z","v":0}
```
### Resend DataFlow event over PubSub to ensure message is received.
```json
{"name":"customs-server","hostname":"stomlinson-30887","pid":60809,"level":30,"op":"dataflow.message","count":1,"id":"512790000031481","data":"{\"severity\":\"info\",\"id\":\"250f4821-6920-4832-bfe3-0a7a1bf1c1a6\",\"summary\":\"test sms send account id rate violation, scotty@mozilla.com\",\"category\":\"customs\",\"timestamp\":\"1970-01-01T00:00:00.000Z\",\"metadata\":[{\"key\":\"customs_category\",\"value\":\"rl_sms_accountid\"},{\"key\":\"threshold\",\"value\":\"5\"},{\"key\":\"count\",\"value\":\"5\"},{\"key\":\"notify_merge\",\"value\":\"rl_sms_accountid\"},{\"key\":\"accountid\",\"value\":\"scotty@mozilla.com\"},{\"key\":\"masked_summary\",\"value\":\"test sms send account id rate violation, <<masked>>\"}]}","attributes":{},"msg":"","time":"2019-04-10T11:12:48.897Z","v":0}
```
### email address is blocked, so `dataflow.block.expected` is logged
```json
{"name":"customs-server","hostname":"stomlinson-30887","pid":60809,"level":30,"customs_category":"rl_sms_accountid","threshold":"5","count":"5","notify_merge":"rl_sms_accountid","accountid":"scotty@mozilla.com","masked_summary":"test sms send account id rate violation, <<masked>>","op":"dataflow.block.expected","msg":"","time":"2019-04-10T11:12:48.899Z","v":0}
```

@mozilla/fxa-devs - r?

File: packages/fxa-customs-server/lib/dataflow.js [hunk 1]
```diff
@@ -54,24 +87,154 @@ module.exports = function (config, log) {
...
+    }
+
+    const dataflowBlock = {}
+
+    parsedBody.metadata.forEach(({ key, value }) => {
+      if (typeof key === 'undefined') {
+        throw new Error('missing metadata key')
+      }
+      if (typeof value === 'undefined') {
+        throw new Error('missing metadata value')
```
Comment (vbudhram) [inline packages/fxa-customs-server/lib/dataflow.js]: We could append the key that is missing the value here for better debugging.

File: packages/fxa-customs-server/lib/dataflow.js [hunk 2]
```diff
@@ -54,24 +87,154 @@ module.exports = function (config, log) {
...
+  }
+
+
+  /**
+   * Check for unexpected DataFlow blocks.
+   * Logs whenever an unexpected block is found.
+   *
+   * @param {Object} dataflowBlock
+   */
+  async function checkForUnexpectedDataflowBlock(dataflowBlock) {
```
Comment (vbudhram) [inline packages/fxa-customs-server/lib/dataflow.js]: nit: Are we expecting other types of blocks in the `dataflow.js` file? If no, I think we could drop `Dataflow` in the function name here.

File: packages/fxa-customs-server/lib/dataflow.js [hunk 3]
```diff
@@ -54,24 +87,154 @@ module.exports = function (config, log) {
...
+    if (! isBlockUnexpected) {
+      log.error({
+        op: 'dataflow.customs_category.unknown',
+        customs_category: category
+      })
+      return
+    }
+
+    if (await isBlockUnexpected(dataflowBlock)) {
+      log.info(Object.assign({}, dataflowBlock, {
```
Comment (vbudhram) [inline packages/fxa-customs-server/lib/dataflow.js]: It is not entirely clear to me what is in a `dataflowBlock` object? Should we amend comment to include/link to documentation?

File: packages/fxa-customs-server/test/local/dataflow_tests.js [hunk 4]
```diff
@@ -1,179 +1,768 @@
...
+      enabled: true,
+      gcpPubSub: {
+        projectId: 'foo',
+        subscriptionName: 'bar'
+      }
     }
+  }
+}
+
+function test(testName, testFunction) {
```
Comment (vbudhram) [inline packages/fxa-customs-server/test/local/dataflow_tests.js]: Tap is one of the reasons working in custom-server is a little more difficult, this makes it slightly more tolerable 👍🏽. But yea, this reminds me that we should eventually standardized with other repos and use mocha.

**Comment by shane-tomlinson** [comment]
@vbudhram - made the requested changes, mind another r?

**Pull request merged**